presto-parquet/src/main/java/io/prestosql/parquet/reader/TimestampColumnReader.java

```diff
@@ -40,13 +47,22 @@ public TimestampColumnReader(RichColumnDescriptor descriptor, DateTimeZone timeZ
     protected void readValue(BlockBuilder blockBuilder, Type type)
     {
         if (definitionLevel == columnDescriptor.getMaxDefinitionLevel()) {
-            long utcMillis = getTimestampMillis(valuesReader.readBytes());
             if (type instanceof TimestampWithTimeZoneType) {
-                type.writeLong(blockBuilder, packDateTimeWithZone(utcMillis, UTC_KEY));
+                int precision = ((TimestampWithTimeZoneType) type).getPrecision();
+                DecodedTimestamp decodedTimestamp = decode(valuesReader.readBytes());
+                if (precision <= TimestampWithTimeZoneType.MAX_SHORT_PRECISION) {
+                    long utcMillis = decodedTimestamp.getEpochSeconds() * MILLISECONDS_PER_SECOND + decodedTimestamp.getNanosOfSecond() / NANOSECONDS_PER_MILLISECOND;
+                    type.writeLong(blockBuilder, packDateTimeWithZone(utcMillis, UTC_KEY));
+                }
+                else {
+                    type.writeObject(blockBuilder, LongTimestampWithTimeZone.fromEpochMillisAndFraction(
+                                    decodedTimestamp.getEpochSeconds() * MILLISECONDS_PER_SECOND + decodedTimestamp.getNanosOfSecond() / NANOSECONDS_PER_MILLISECOND,
+                                    decodedTimestamp.getNanosOfSecond() % NANOSECONDS_PER_MILLISECOND * PICOSECONDS_PER_NANOSECOND, UTC_KEY));
+                }
             }
             else {
-                utcMillis = timeZone.convertUTCToLocal(utcMillis);
-                type.writeLong(blockBuilder, utcMillis * MICROSECONDS_PER_MILLISECOND);
+                PrestoTimestampEncoder<?> prestoTimestampEncoder = createTimestampEncoder((TimestampType) type, timeZone);
```

Comment:
We *should* do a refactor where the type is provided at the construction time (it's const), as `createTimestampEncoder` is now called per row


add a TODO comment & respective issues